### PR TITLE
Add WAITPKG checks, add support for TPAUSE in ThreadPool spin

### PR DESCRIPTION
**Pull request by kleiti**
### Description
Adding necessary runtime checks for WAITPKG:
onnxruntime/core/common/cpuid_info.cc
onnxruntime/core/common/cpuid_info.h

Added a spin function with TPAUSE, use it in WorkerLoop if HW has WAITPKG support. Defining the control bit as 0, since this will provide better power savings, but also improve performance of other SMT thread on the same core. See https://www.felixcloutier.com/x86/tpause.
include/onnxruntime/core/common/spin_pause.h
include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

### Motivation and Context
Change enables TPAUSE support on Alder Lake and newer Intel microarchitectures in the context of Eigen-based threadpool implementation. This allows the CPU to enter light-weight C-state while in a spin, saving power and in this case, improving performance by 3-4%.
See section 'Active Spins in Applications' in
https://cdrdv2-public.intel.com/685865/211112_Hybrid_WP_2_Developing_v1.2.pdf

### Result
Tested with onnxruntime_perf_test.exe on a dynamically quantized model (which I can not share - however these changes are not model specific), running 
`onnxruntime_perf_test.exe -I -m times -r 10000 -o 2 model.onnx`

As to optimal TPAUSE cycle count, testing shows that ~2000 cycles provides the best results. First 2 graphs were generated from onnxruntime_perf_test.exe output, power data collected using Intel Power Gadget on ADL i9-12700H.
image
image
image

### CODE REVIEW

**Comment by kleiti** [comment]
@microsoft-github-policy-service agree company="Intel"

**Comment by kleiti** [comment]
@tlh20 Since this PR touches your original implementation, proposing for you to review it.

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h [hunk 1]
```diff
@@ -1554,8 +1559,14 @@ class ThreadPoolTempl : public onnxruntime::concurrency::ExtendedThreadPoolInter
...
-          onnxruntime::concurrency::SpinPause();
-        }
+		  
+          if (tpause) {
+            onnxruntime::concurrency::SpinTPAUSE();  
+          }
+          else {
+            onnxruntime::concurrency::SpinPause();
+          }
+        }
```
Comment (tlh20) [inline include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h:1569]: We use SpinPause in a few other places -- probably the barrier and lock implementations.

Are we likely to benefit from using SpinTPAUSE in those cases also?

Unless there's a good reason to keep the conditional here, I'd tend to push it into a common function in onnxruntime::concurrency where we would select the appropriate pause operation for the CPU we are running on

File: onnxruntime/core/common/cpuid_info.cc [hunk 2]
```diff
@@ -83,6 +83,15 @@ static inline int XGETBV() {
 #endif
 }
 
+int filter(uint32_t code) {
```
Comment (tlh20) [inline onnxruntime/core/common/cpuid_info.cc:86]:
```suggestion
// Exception filter used when testing operations that we can detect only by attempting to run them
int filter(uint32_t code) {
```

File: include/onnxruntime/core/common/spin_pause.h [hunk 3]
```diff
@@ -23,6 +23,12 @@ inline void SpinPause() {
 #endif
 }
 
+inline void SpinTPAUSE() {
+#if defined(_M_AMD64) || defined(__x86_64__)
+  _tpause(0x0, __rdtsc() + 2000);
```
Comment (tlh20) [inline include/onnxruntime/core/common/spin_pause.h:28]: Maybe add a local constant spin_delay_cycles=2000 (or using the appropriate unit) to be explicit what the number 2000 indicates

**Comment by kleiti** [comment]
@tlh20 Thank you for the review and apologies for the delay in getting back to you. I'm in the middle of testing this PR with some other models (mobilenetv2-12.onnx, for example) so we can get a better understanding of how performance is affected. I'll post those results in the next few days and will look at your suggestions then.

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h (hunk 1 above)
Comment (kleiti) [reply include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h:1569]: As for the SpinPause being used in other places - I did not see meaningful performance gains when I tested with TPAUSE.

Could you be more specific on the second request? I thought of moving the CPUInfo call to spinpause.h, but even if it's in the same namespace, the function is not available there.

File: include/onnxruntime/core/common/spin_pause.h (hunk 3 above)
Comment (kleiti) [reply include/onnxruntime/core/common/spin_pause.h:28]: Sure, that'll make it clearer. I'll make the change.

**Comment by kleiti** [comment]
@tlh20 This change could provide both performance and particularly power benefits to CPU inferencing. For the below results, I've used MobileNetV2 model
`onnxruntime_perf_test.exe` -I -m times -r 10000 -o 2 `mobilenetv2-12.onnx` on a 12th Gen Intel i7-12700H CPU.

image

image

image

Data (with some additional information) in table format:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Intel/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Intel/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">


mm_pause | Average |   | tpause(2000) | Average
-- | -- | -- | -- | --
Cumulative Processor Energy_0   (Joules) | 7283.41698 |   | Cumulative Processor Energy_0 (Joules) | 6052.896
Cumulative Processor Energy_0 (mWh) | 2023.17138 |   | Cumulative Processor Energy_0 (mWh) | 1681.36
Average Processor Power_0 (Watt) | 73.3675714 |   | Average Processor Power_0 (Watt) | 61.72873
  |   |   |   |  
Cumulative IA Energy_0 (Joules) | 7184.39187 |   | Cumulative IA Energy_0 (Joules) | 5959.524
Cumulative IA Energy_0 (mWh) | 1995.66441 |   | Cumulative IA Energy_0 (mWh) | 1655.423
Average IA Power_0 (Watt) | 72.3702642 |   | Average IA Power_0 (Watt) | 60.77651



</body>

</html>

**Comment by kleiti** [comment]
SpinPause is being added to DML EP now as well.
https://github.com/microsoft/onnxruntime/blob/6bd6d879a351e2ed4df84357f826d3df60539f34/onnxruntime/core/providers/dml/DmlExecutionProvider/src/GpuEvent.h#L33

image

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h (hunk 1 above)
Comment (kleiti) [reply include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h]: I've moved the TPAUSE check to spinpause.h and the selection (mm_pause or tpause) now happens in SpinPause(). This allows me to leave EigenNonBlockingThreadPool.h untouched and will let the DML EP change that I mention elsewhere in this PR to take advantage of this optimization as well.

**Pull request closed without merging**